File: core/core-api/src/main/java/com/coffee/gu/api/controller/ApiControllerAdvice.java
package com.coffee.gu.api.controller;

import com.coffee.gu.CoreException;
import com.coffee.gu.ErrorType;
import com.coffee.gu.response.ApiResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class ApiControllerAdvice {
    private static final Logger log = LoggerFactory.getLogger(ApiControllerAdvice.class);

    @ExceptionHandler(CoreException.class)
    ResponseEntity<ApiResponse<?>> handleCoreException(CoreException e) {
        switch (e.getErrorType().getErrorLogLevel()) {
            case ERROR -> log.error("com.coffee.gu.CoreException : {}", e.getMessage(), e);
            case WARN -> log.warn("com.coffee.gu.CoreException : {}", e.getMessage(), e);
            default -> log.info("com.coffee.gu.CoreException : {}", e.getMessage(), e);
        }
        return new ResponseEntity<>(ApiResponse.error(e.getErrorType(), e.getMessage()), HttpStatusCode.valueOf(e.getErrorType().getStatus()));
    }

    @ExceptionHandler(Exception.class)
    ResponseEntity<ApiResponse<?>> handleException(Exception e) {
        log.error("Exception : {}", e.getMessage(), e);
        return new ResponseEntity<>(ApiResponse.error(ErrorType.DEFAULT_ERROR, e.getMessage()), HttpStatusCode.valueOf(ErrorType.DEFAULT_ERROR.getStatus()));
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/api/controller/ApiControllerAdvice.kt
package com.coffee.gu.api.controller

import com.coffee.gu.CoreException
import com.coffee.gu.ErrorLogLevel
import com.coffee.gu.ErrorType
import com.coffee.gu.response.ApiResponse
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ApiControllerAdvice {
    companion object {
        private val log = LoggerFactory.getLogger(ApiControllerAdvice::class.java)
    }

    @ExceptionHandler(CoreException::class)
    fun handleCoreException(e: CoreException): ResponseEntity<ApiResponse<*>> {
        when (e.errorType.errorLogLevel) {
            ErrorLogLevel.ERROR -> log.error("com.coffee.gu.CoreException : {}", e.message, e)
            ErrorLogLevel.WARN -> log.warn("com.coffee.gu.CoreException : {}", e.message, e)
            else -> log.info("com.coffee.gu.CoreException : {}", e.message, e)
        }
        return ResponseEntity(ApiResponse.error<Any>(e.errorType, e.message), HttpStatusCode.valueOf(e.errorType.status))
    }

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ApiResponse<*>> {
        log.error("Exception : {}", e.message, e)
        return ResponseEntity(ApiResponse.error<Any>(ErrorType.DEFAULT_ERROR, e.message), HttpStatusCode.valueOf(ErrorType.DEFAULT_ERROR.status))
    }
}
